### ISSUE TITLE
Prometheus add-on test plan

### ISSUE THREAD

**Issue by kleimkuhler** [linkerd/linkerd2#5041]
This is a continuation of #4386 and tracks the test plan for the Prometheus
add-on.

This was originally part of the test plan laid out in #4386, but was ultimately
removed as it did not ship with 2.8

- [x] Bring your own Prometheus use-case
    - Install Prometheus with scrape config.
    - Install Linkerd with Prometheus disabled & cofiguring global.promethusUrl to existing Prometheus
    - Linkerd Check Tooling
    - Install emojivoto/books app and test for usual metrics through dashboard and CLI
    - Upgrade the control-plane, without having to pass configuration again.

- [x] Linkerd with long-term storage use-case
    - Install Cortex or any other long-term metrics storage solution which follows the Prometheus query API like Thanos, etc.
    - Install Linkerd with Prometheus enabled, with the below configuration for the above linked cortex manifest.
    - Dashboards and CLI should work, as the metrics are directly being queried from long-term storage. Restarting Prometheus pod would cause the current metrics to disappear but the previous metrics should be present as they are from the long term storage. (can be seen in grafana)
    - Same should work well with upgrades, without causing any configuration loss.

```
global:
  prometheusUrl: http://cortex-service.default.svc.cluster.local:9009/api/prom
prometheus:
  remote_write:
    - url: http://cortex-service.default.svc.cluster.local:9009/api/prom/push
```

- [x] With full Prometheus configuration
    - Install Linkerd with full Prometheus configuration, and checking every configuration is applied correctly with Prometheus i.e remote-writes, scrape-configs, args, etc.
    - Upgrades should work correctly, along with linkerd-config-addons having correct configuration.

**Comment by kleimkuhler** [comment]
For testing the "bring your own Prometheus" use-case, I am disabling the bundled Linkerd Prometheus and using one that I have already installed on the cluster.

To install a separate Prometheus, I use the `prometheus.yaml` resource file from below. It installs Prometheus in the `default` namespace. It uses the same roles and scrape config that the Linkerd Prometheus uses.

I then install Linkerd with the `addon-config.yaml` resource file from below. It sets the Prometheus URL that the control plane uses to the one that is already installed and disables the add-on:

```
linkerd install --addon-config addon-config.yaml
```

After installation, `linkerd check` should complete successfully and the control plane should be able to talk to Prometheus:

```
linkerd-api
-----------
..
√ [prometheus] control plane can talk to Prometheus
..
```

`prometheus.yaml`:

```
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: prometheus
rules:
- apiGroups: [""]
  resources: ["nodes", "nodes/proxy", "pods"]
  verbs: ["get", "list", "watch"]
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: prometheus
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: prometheus
subjects:
- kind: ServiceAccount
  name: prometheus
  namespace: default
---
kind: ServiceAccount
apiVersion: v1
metadata:
  name: prometheus
  namespace: default
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: prometheus-config
  namespace: default
data:
  prometheus.yml: |-
    global:
      evaluation_interval: 10s
      scrape_interval: 10s
      scrape_timeout: 10s

    rule_files:
    - /etc/prometheus/*_rules.yml
    - /etc/prometheus/*_rules.yaml

    scrape_configs:
    - job_name: 'prometheus'
      static_configs:
      - targets: ['localhost:9090']

    - job_name: 'grafana'
      kubernetes_sd_configs:
      - role: pod
        namespaces:
          names: ['linkerd']
      relabel_configs:
      - source_labels:
        - __meta_kubernetes_pod_container_name
        action: keep
        regex: ^grafana$

    #  Required for: https://grafana.com/grafana/dashboards/315
    - job_name: 'kubernetes-nodes-cadvisor'
      scheme: https
      tls_config:
        ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
        insecure_skip_verify: true
      bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
      kubernetes_sd_configs:
      - role: node
      relabel_configs:
      - action: labelmap
        regex: __meta_kubernetes_node_label_(.+)
      - target_label: __address__
        replacement: kubernetes.default.svc:443
      - source_labels: [__meta_kubernetes_node_name]
        regex: (.+)
        target_label: __metrics_path__
        replacement: /api/v1/nodes/$1/proxy/metrics/cadvisor
      metric_relabel_configs:
      - source_labels: [__name__]
        regex: '(container|machine)_(cpu|memory|network|fs)_(.+)'
        action: keep
      - source_labels: [__name__]
        regex: 'container_memory_failures_total' # unneeded large metric
        action: drop

    - job_name: 'linkerd-controller'
      kubernetes_sd_configs:
      - role: pod
        namespaces:
          names: ['linkerd']
      relabel_configs:
      - source_labels:
        - __meta_kubernetes_pod_label_linkerd_io_control_plane_component
        - __meta_kubernetes_pod_container_port_name
        action: keep
        regex: (.*);admin-http$
      - source_labels: [__meta_kubernetes_pod_container_name]
        action: replace
        target_label: component

    - job_name: 'linkerd-service-mirror'
      kubernetes_sd_configs:
      - role: pod
      relabel_configs:
      - source_labels:
        - __meta_kubernetes_pod_label_linkerd_io_control_plane_component
        - __meta_kubernetes_pod_container_port_name
        action: keep
        regex: linkerd-service-mirror;admin-http$
      - source_labels: [__meta_kubernetes_pod_container_name]
        action: replace
        target_label: component

    - job_name: 'linkerd-proxy'
      kubernetes_sd_configs:
      - role: pod
      relabel_configs:
      - source_labels:
        - __meta_kubernetes_pod_container_name
        - __meta_kubernetes_pod_container_port_name
        - __meta_kubernetes_pod_label_linkerd_io_control_plane_ns
        action: keep
        regex: ^linkerd-proxy;linkerd-admin;linkerd$
      - source_labels: [__meta_kubernetes_namespace]
        action: replace
        target_label: namespace
      - source_labels: [__meta_kubernetes_pod_name]
        action: replace
        target_label: pod
      # special case k8s' "job" label, to not interfere with prometheus' "job"
      # label
      # __meta_kubernetes_pod_label_linkerd_io_proxy_job=foo =>
      # k8s_job=foo
      - source_labels: [__meta_kubernetes_pod_label_linkerd_io_proxy_job]
        action: replace
        target_label: k8s_job
      # drop __meta_kubernetes_pod_label_linkerd_io_proxy_job
      - action: labeldrop
        regex: __meta_kubernetes_pod_label_linkerd_io_proxy_job
      # __meta_kubernetes_pod_label_linkerd_io_proxy_deployment=foo =>
      # deployment=foo
      - action: labelmap
        regex: __meta_kubernetes_pod_label_linkerd_io_proxy_(.+)
      # drop all labels that we just made copies of in the previous labelmap
      - action: labeldrop
        regex: __meta_kubernetes_pod_label_linkerd_io_proxy_(.+)
      # __meta_kubernetes_pod_label_linkerd_io_foo=bar =>
      # foo=bar
      - action: labelmap
        regex: __meta_kubernetes_pod_label_linkerd_io_(.+)
      # Copy all pod labels to tmp labels
      - action: labelmap
        regex: __meta_kubernetes_pod_label_(.+)
        replacement: __tmp_pod_label_$1
      # Take `linkerd_io_` prefixed labels and copy them without the prefix
      - action: labelmap
        regex: __tmp_pod_label_linkerd_io_(.+)
        replacement:  __tmp_pod_label_$1
      # Drop the `linkerd_io_` originals
      - action: labeldrop
        regex: __tmp_pod_label_linkerd_io_(.+)
      # Copy tmp labels into real labels
      - action: labelmap
        regex: __tmp_pod_label_(.+)
---
kind: Service
apiVersion: v1
metadata:
  name: prometheus
  namespace: default
spec:
  type: ClusterIP
  selector:
    app: prometheus
  ports:
  - name: admin-http
    port: 9090
    targetPort: 9090
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: prometheus
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: prometheus
  template:
    metadata:
      labels:
        app: prometheus
    spec:
      nodeSelector:
        beta.kubernetes.io/os: linux
      securityContext:
        fsGroup: 65534
      containers:
      - args:
        - --config.file=/etc/prometheus/prometheus.yml
        - --log.level=info
        - --storage.tsdb.path=/data
        - --storage.tsdb.retention.time=6h
        image: prom/prometheus:v2.19.3
        imagePullPolicy: IfNotPresent
        livenessProbe:
          httpGet:
            path: /-/healthy
            port: 9090
          initialDelaySeconds: 30
          timeoutSeconds: 30
        name: prometheus
        ports:
        - containerPort: 9090
          name: admin-http
        readinessProbe:
          httpGet:
            path: /-/ready
            port: 9090
          initialDelaySeconds: 30
          timeoutSeconds: 30
        securityContext:
          runAsNonRoot: true
          runAsUser: 65534
          runAsGroup: 65534
        volumeMounts:
        - mountPath: /data
          name: data
        - mountPath: /etc/prometheus/prometheus.yml
          name: prometheus-config
          subPath: prometheus.yml
          readOnly: true
      serviceAccountName: prometheus
      volumes:
      - name: data
        emptyDir: {}
      - configMap:
          name: prometheus-config
        name: prometheus-config
```

`addon-config.yaml`:

```
global:
  prometheusUrl: http://prometheus.default.svc.cluster.local:9090
prometheus:
  enabled: false
```

**Comment by kleimkuhler** [comment]
I have confirmed this works with Cortex as well.

I have tested both using `linkerd-prometheus` and BYO Prometheus.

### `linkerd-prometheus`

`addon-config.yaml`:

```
global:
  prometheusUrl: http://cortex-service.default.svc.cluster.local:9009/api/prom
prometheus:
  remote_write:
    - url: http://cortex-service.default.svc.cluster.local:9009/api/prom/push
```

### BYO Prometheus

`addon-config.yaml`
```
global:
  prometheusUrl: http://cortex-service.default.svc.cluster.local:9009/api/prom
prometheus:
  enabled: false
```

This Prometheus differs from the one above because I have added the `remote_write` field in the config: [prometheus.yaml](https://gist.github.com/kleimkuhler/7de84e81b60e52354d09e218c385f4c4)

### Cortex and Cassandra

I am creating a Cortex deployment with @Pothulapati's gist [here](https://gist.github.com/Pothulapati/3b42bebabf38863a5e4674ffd925fe4d).

### Steps

I installed Linkerd with the `addon-config.yaml` and inject Cortex/Cassandra deployments and Prometheus deployment if I am bringing my own.

I can delete `linkerd-prometheus` or `prometheus.yaml` and observe that the metrics stay in Granfana.

**Comment by kleimkuhler** [comment]
All cases have been covered with `edge-20.10.1`.

**Comment by kleimkuhler** [comment]
All cases have again been covered with `edge-20.10.3`.

I originally encountered some issues with metrics not displaying when expected, but I've found this is some general flakiness with the `cortex.yaml` I've been using for testing. I've been able to reproduce the same issues with the `edge-20.10.1` and usually it just takes some restarts of the `cortex` and `cassandra` deployments _if_ they don't work right from the start.
